cpp/src/strings/regex/regcomp.cpp

```diff
@@ -239,82 +240,69 @@ class regex_parser {
-      if (!is_quoted && chr == ']' && count_char > 1) break;
+      if (!is_quoted && chr == ']' && count_char > 1) { break; }  // done
       if (!is_quoted && chr == '-') {
-        if (cls.empty()) {
-          // malformed '[]': TODO assert or exception?
-          return 0;
-        }
+        if (literals.empty()) { return 0; }  // malformed '[]'
         std::tie(is_quoted, chr) = next_char();
-        if ((!is_quoted && chr == ']') || chr == 0) {
-          // malformed '[]': TODO assert or exception?
-          return 0;
-        }
-        cls.back() = chr;
+        if ((!is_quoted && chr == ']') || chr == 0) { return 0; }  // malformed '[]'
+        literals.back() = chr;
       } else {
-        cls.push_back(chr);
-        cls.push_back(chr);
+        literals.push_back(chr);
+        literals.push_back(chr);
       }
       std::tie(is_quoted, chr) = next_char();
     }
 
-    /* sort on span start */
-    for (std::size_t p = 0; p < cls.size(); p += 2)
-      for (std::size_t np = p + 2; np < cls.size(); np += 2)
-        if (cls[np] < cls[p]) {
-          auto c      = cls[np];
-          cls[np]     = cls[p];
-          cls[p]      = c;
-          c           = cls[np + 1];
-          cls[np + 1] = cls[p + 1];
-          cls[p + 1]  = c;
-        }
-
-    /* merge spans */
-    reclass yycls{builtins};
-    if (cls.size() >= 2) {
-      int np        = 0;
-      std::size_t p = 0;
-      yycls.literals += cls[p++];
-      yycls.literals += cls[p++];
-      for (; p < cls.size(); p += 2) {
-        /* overlapping or adjacent ranges? */
-        if (cls[p] <= yycls.literals[np + 1] + 1) {
-          if (cls[p + 1] >= yycls.literals[np + 1])
-            yycls.literals.replace(np + 1, 1, 1, cls[p + 1]); /* coalesce */
-        } else {
-          np += 2;
-          yycls.literals += cls[p];
-          yycls.literals += cls[p + 1];
-        }
+    // transform pairs of literals to ranges
+    std::vector<reclass_range> ranges(literals.size() / 2);
+    auto const counter = thrust::make_counting_iterator(0);
+    std::transform(counter, counter + ranges.size(), ranges.begin(), [&literals](auto idx) {
+      return reclass_range{literals[idx * 2], literals[idx * 2 + 1]};
+    });
+    // sort the ranges to help with detecting overlapping entries
+    std::sort(ranges.begin(), ranges.end(), [](auto l, auto r) {
+      return l.first == r.first ? l.last < r.last : l.first < r.first;
+    });
+    // combine overlapping entries: [a-f][c-g] => [a-g]
+    for (auto itr = ranges.begin() + static_cast<int>(!ranges.empty()); itr < ranges.end(); ++itr) {
+      auto const prev = *(itr - 1);
+      if (itr->first <= prev.last + 1) {
+        // if these 2 ranges intersect, expand the current one
+        *itr = reclass_range{prev.first, std::max(prev.last, itr->last)};
       }
     }
-    _cclass_id = _prog.add_class(yycls);
+    // remove any duplicates
+    std::reverse(ranges.begin(), ranges.end());  // moves larger overlaps forward
```

Comment:
Seems an unfortunate thing to have to do to keep `std::unique` from tossing out the wrong entry.

1. Is this a performance issue? Has it been measured to compare to the original method?
2. Would it make sense to use a reverse iterator instead?

Reply:
There is not performance concern. I don't expect there to be more than half dozen entries here. I can look at the reverse iterator. If it still returns the same `end` iterator then that would be cool.

Reply:
https://godbolt.org/z/zqG7xTvEG